### PR TITLE
Feature/hash failure domain naming

### PR DESCRIPTION
**Pull request by rejoshed**
*Issue #, if available:*

*Description of changes:*
Meta naming for failure domain custom resource objects currently follows `failureDomainName` + `-` + `clusterName`.

This was to work around EKS-A's decision to put all clusters in a single namespace.

Unfortunately k8s object names can only be 64 characters long and this leads to the concern that CAPC will break for any combination of names that leads to >64 character failure domain naming.

This uses a hash of the failure domain name and the cluster name to predictably generate a failure domain's meta name.

Also should close https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/issues/113#issue-1259386590

*Testing performed:*
All unit tests passing locally.

Ran e2e PR-Blocking tests successfully.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1beta2/cloudstackmachine_types.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ type CloudStackMachineSpec struct {
 
 	// Optional affinitygroupids for deployVirtualMachine
 	// +optional
-	AffinityGroupIDs []string `json:"affinitygroupids,omitempty"`
+	AffinityGroupIDs []string `json:"affinityGroupIDs,omitempty"`
```
Comment (rejoshed) [inline api/v1beta2/cloudstackmachine_types.go:66]: Just a fixup for JSON tagging consistency.

File: controllers/cloudstackfailuredomain_controller.go [hunk 2]
```diff
@@ -84,17 +84,14 @@ func (r *CloudStackFailureDomainReconciliationRunner) Reconcile() (retRes ctrl.R
...
-	}
 
 	// Check if the passed network was an isolated network or the network was missing. In either case, create a
 	// CloudStackIsolatedNetwork to manage the many intricacies and wait until CloudStackIsolatedNetwork is ready.
 	if r.ReconciliationSubject.Spec.Zone.Network.ID == "" ||
 		r.ReconciliationSubject.Spec.Zone.Network.Type == infrav1.NetworkTypeIsolated {
 		netName := r.ReconciliationSubject.Spec.Zone.Network.Name
-		if res, err := r.GenerateIsolatedNetwork(netName)(); r.ShouldReturn(res, err) {
+		if res, err := r.GenerateIsolatedNetwork(
+			netName, func() string { return r.ReconciliationSubject.Spec.Name })(); r.ShouldReturn(res, err) {
```
Comment (rejoshed) [inline controllers/cloudstackfailuredomain_controller.go:94]: Need to pass failure domain name in here.

File: config/manager/manager.yaml [hunk 3]
```diff
@@ -15,8 +15,8 @@ spec:
       labels:
         control-plane: capc-controller-manager
     spec:
-      securityContext:
-        runAsNonRoot: true
+      # securityContext:
```
Comment (wongni) [inline config/manager/manager.yaml:18]: Please remote comments

File: test/e2e/config/cloudstack.yaml [hunk 4]
```diff
@@ -113,7 +113,7 @@ variables:
   NODE_DRAIN_TIMEOUT: "60s"
 
   CLOUDSTACK_FD1_NAME: "fd1"
-  CLOUDSTACK_FD1_SECRET_NAME: "secret1"
+  CLOUDSTACK_FD1_SECRET_NAME: "myendpoint"
```
Comment (wongni) [inline test/e2e/config/cloudstack.yaml:116]: Please revert this back to secret1

File: api/v1beta2/cloudstackmachine_types.go (hunk 1 above)
Comment (maxdrib) [inline api/v1beta2/cloudstackmachine_types.go:66]: I don't think we should change this now. EKS-A is already and other customers may also be using this field and changing the serialization/deserialization patterns feels risky, with no gain. I'm inclined to keep these as they are and fix later on in an intentional way

Unless this is strictly for the v1beta2 types, and not v1beta1? Same for the other changes involving json

**Review by wongni — APPROVED** [review]

File: controllers/utils/failuredomains.go [hunk 5]
```diff
@@ -73,22 +71,30 @@ func (r *ReconciliationRunner) GetFailureDomains(fds *infrav1.CloudStackFailureD
 	}
 }
 
+// GetFailureDomainByName gets a singel FailureDomain by name and requeues if it's not found.
```
Comment (maxdrib) [inline controllers/utils/failuredomains.go:74]:
```suggestion
// GetFailureDomainByName gets a single FailureDomain by name and requeues if it's not found.
```

File: api/v1beta2/cloudstackisolatednetwork_types.go [hunk 6]
```diff
@@ -41,7 +41,7 @@ type CloudStackIsolatedNetworkSpec struct {
 
 	//+k8s:conversion-gen=false
 	// FailureDomain -- the FailureDomain the network is placed in.
```
Comment (maxdrib) [inline api/v1beta2/cloudstackisolatednetwork_types.go:43]: update comment

**Review by maxdrib — COMMENTED** [review]
Why are all these test data files deleted?

**Comment by wongni** [comment]
> Why are all these test data files deleted?

They are generated files via `kustomize` during executing run-e2e make target

File: api/v1beta2/cloudstackmachine_types.go (hunk 1 above)
Comment (rejoshed) [reply api/v1beta2/cloudstackmachine_types.go:66]: Indeed, this is only on v1beta2 types.

That's why I wanted to jamb this into this release. It won't be easy to do outside of an API version bump.

File: api/v1beta2/cloudstackisolatednetwork_types.go (hunk 6 above)
Comment (rejoshed) [reply api/v1beta2/cloudstackisolatednetwork_types.go:43]: Name changed.

File: controllers/utils/failuredomains.go (hunk 5 above)
Comment (rejoshed) [reply controllers/utils/failuredomains.go:74]: Changed.

**Review by maxdrib — APPROVED** [review]

**Pull request merged**